_posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md

```diff
@@ -0,0 +1,133 @@
+        |> make_an_attack()
+        |> set_as_last_shooter()
+    else
+      ignore()
+    end
+end
+```
+
+Or even pattern matching with guard clause
+
+```elixir
+def maybe_attack(player, last_shooter) when player == last_shooter do
+    ignore()
+end
+
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+But it is possible to combine the pattern matching we already used in the guard clause solution with the power of immutability to come up with an even more alchemistic solution!
+
+```elixir
+def maybe_attack(last_shooter, last_shooter) do
+    ignore()
+end
+
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Wait a second, what have we done here?
+
+In the first `maybe_attack` function we are calling our player (the first argument) as last_shooter!
+That makes this function only match if the player is also the last_shooter!
+Thanks to immutability, when we call both arguments by the same name Elixir will check if they are actually the same!
+We could easily call both arguments player or even something like player_is_the_last_shooter. It doesn't matter! The rule is just that if we want to ensure that there is equality, we call both arguments by the same name!
+
+**Ok, it's time to play using our nice little code!**
+
+Let's say we have `player1` and `player2`, `player1` made the most recent attack and therefore is our last_shooter and now `player2` will try to attack!
+
+So we will call the function `maybe_attack(player2, player1)` where `player2` is the player who wants to make an attack and `player1` is our last_shooter
+
+We have two `maybe_attack` functions both with arity 2, so Elixir will try to pattern match from top to bottom, i.e. the first function it will try to match will be 
+
+```elixir
+def maybe_attack(last_shooter, last_shooter) do
+    ignore()
+end
+```
+
+Our first argument is `player2` and Elixir will bind it with `last_shooter = player2` and since our second argument is also a `last_shooter`, Elixir will use the `^` (pin operator) to check if the previous bind is valid for the second argument instead of trying to rebind it.
+
+```elixir
+last_shooter = player2
+ˆlast_shooter = player1
+```
+
+As player2 is different from player1, we will not have a valid pattern matching and therefore the elixir will move on to try to match with the next function!
+
+**Our next match attempt!**
+
+```elixir
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Now the behavior will be different, we are asking Elixir to match two different arguments. That is, to make a bind for each one.
+
+```elixir
+player = player2
+last_shooter = player1
+```
+
+With a valid match, our function will run! Player2 will make an attack and then will be registered as our new last_shooter!
+
+What if player2 tries another attack in a row?
```

Comment:
here, wrap "last_shooter" in backtics "``", same with "maybe_attack" and another mention of "last_shooter" below.